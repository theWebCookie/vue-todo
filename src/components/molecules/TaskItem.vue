<template>
  <li :class="$style.taskItem">
    <InputCheckbox :checked="task.completed" :disabled="false" :id="task.id" />
    <div :class="$style.taskName">
      <label :for="task.id">{{ task.title }}</label>
      <AppButton :customClass="$style.deleteButton" @click="removeTask(task.id)"> X </AppButton>
    </div>
  </li>
</template>

<script>
import AppButton from '../atoms/AppButton.vue';
import InputCheckbox from '../atoms/InputCheckbox.vue';

export default {
  name: 'TaskItem',
  components: {
    InputCheckbox,
    AppButton,
  },
  props: {
    task: {
      type: Object,
      required: true,
    },
  },
  methods: {
    removeTask(taskId) {
      this.$store.dispatch('removeTask', taskId);
    },
  },
};
</script>

<style module lang="scss">
.taskItem {
  display: flex;
  gap: 10px;
  padding: 15px 10px;
  border-bottom: 1px solid #ccc;

  .taskName {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &:hover .deleteButton {
    visibility: visible;
  }
}

.deleteButton {
  visibility: hidden;
}
</style>
