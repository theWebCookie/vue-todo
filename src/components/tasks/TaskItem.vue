<template>
  <li :class="$style.taskItem">
    <input type="checkbox" :id="task.id" :checked="task.completed" @change="toggleTask(task.id)" />
    <div :class="$style.taskName">
      <label :for="task.id">{{ task.title }}</label>
      <button :class="$style.deleteButton" @click="removeTask(task.id)">X</button>
    </div>
  </li>
</template>

<script>
export default {
  name: 'TaskItem',
  props: {
    task: {
      type: Object,
      required: true,
    },
  },
  methods: {
    toggleTask(taskId) {
      this.$store.dispatch('toggleTask', taskId);
    },
    removeTask(taskId) {
      this.$store.dispatch('removeTask', taskId);
    },
  },
};
</script>

<style module lang="scss">
.taskItem {
  display: flex;
  gap: 10px;
  padding: 15px 10px;
  border-bottom: 1px solid #ccc;

  .taskName {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;

    .deleteButton {
      all: unset;
      cursor: pointer;
      visibility: hidden;
    }
  }

  &:hover .deleteButton {
    visibility: visible;
  }
}
</style>
