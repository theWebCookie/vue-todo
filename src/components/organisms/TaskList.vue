<template>
  <ul :class="$style.list">
    <li v-if="!tasks.length" :class="$style.noTasks">No tasks yet</li>
    <TaskItem v-for="task in tasks" :key="task.id" :task="task" />
  </ul>
</template>

<script>
import TaskItem from '../molecules/TaskItem.vue';

export default {
  name: 'TaskList',
  components: {
    TaskItem,
  },
  computed: {
    tasks() {
      return this.$store.getters.filteredTasks;
    },
  },
};
</script>

<style module lang="scss">
.list {
  list-style: none;
  padding: 0;
  margin: 0;
  background-color: #fff;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  box-shadow: 0px 38px 12px 0px rgb(120, 120, 120);
  max-height: 300px;
  overflow-y: auto;

  .noTasks {
    padding: 10px;
    text-align: center;
    color: #b0b0b0;
  }
}
</style>
