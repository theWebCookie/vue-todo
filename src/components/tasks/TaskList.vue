<template>
  <ul :class="$style.list">
    <TaskItem v-for="task in tasks" :key="task.id" :task="task" :toggle-task="toggleTask" />
  </ul>
</template>

<script>
import TaskItem from './TaskItem.vue';

export default {
  name: 'TaskList',
  components: {
    TaskItem,
  },
  props: {
    tasks: {
      type: Array,
      required: true,
    },
  },
  methods: {
    toggleTask(taskId) {
      const task = this.tasks.find((t) => t.id === taskId);
      if (task) {
        task.completed = !task.completed;
        console.log(this.tasks);
      }
    },
  },
};
</script>

<style module>
.list {
  list-style: none;
  padding: 0;
  margin: 0;
  background-color: #fff;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  box-shadow: 0px 38px 12px 0px rgb(120, 120, 120);
}
</style>
