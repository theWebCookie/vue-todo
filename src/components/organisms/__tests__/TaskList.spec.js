import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, jest } from '@jest/globals';
import { createStore } from 'vuex';
import TaskList from '../TaskList.vue';
import TaskItem from '../../molecules/TaskItem.vue';

const styles = {
  list: 'list-class',
  noTasks: 'no-tasks-class',
};

describe('TaskList.vue', () => {
  let store;
  let getters;

  it('renders correctly with no tasks', () => {
    getters = {
      filteredTasks: () => [],
    };

    store = createStore({
      getters,
    });

    const wrapper = shallowMount(TaskList, {
      global: {
        plugins: [store],
        mocks: {
          $style: styles,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders correctly with tasks', () => {
    getters = {
      filteredTasks: () => [
        {
          id: 1,
          text: 'Task 1',
          completed: false,
        },
        {
          id: 2,
          text: 'Task 2',
          completed: true,
        },
      ],
    };

    store = createStore({
      getters,
    });

    const wrapper = shallowMount(TaskList, {
      global: {
        plugins: [store],
        mocks: {
          $style: styles,
        },
        stubs: {
          TaskItem,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findAllComponents(TaskItem)).toHaveLength(2);
  });

  it('initializes sortable on mount', () => {
    const spy = jest.spyOn(TaskList.methods, 'initializeDragAndDrop');
    shallowMount(TaskList, {
      global: {
        plugins: [store],
        mocks: {
          $style: styles,
        },
      },
    });
    expect(spy).toHaveBeenCalled();
  });
});
