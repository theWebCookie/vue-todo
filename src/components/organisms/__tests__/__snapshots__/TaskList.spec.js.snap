// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList.vue renders correctly with no tasks 1`] = `
<ul
  class="list"
>
  <li
    class="noTasks"
    draggable="false"
  >
    No tasks yet
  </li>
  
  
</ul>
`;

exports[`TaskList.vue renders correctly with tasks 1`] = `
<ul
  class="list"
>
  <!--v-if-->
  
  <li
    class="taskItem"
    draggable="true"
  >
    <label
      class="taskLabel"
      for="1"
    >
      <input-checkbox-stub
        checked="false"
        disabled="false"
        id="1"
      />
      <div
        class="taskDescription"
      >
        <span
          class="taskTitle"
        />
        <app-button-stub
          customclass="deleteButton"
          isactive="false"
        />
      </div>
    </label>
  </li>
  <li
    class="taskItem taskCompleted"
    draggable="true"
  >
    <label
      class="taskLabel"
      for="2"
    >
      <input-checkbox-stub
        checked="true"
        disabled="false"
        id="2"
      />
      <div
        class="taskDescription"
      >
        <span
          class="taskTitle"
        />
        <app-button-stub
          customclass="deleteButton"
          isactive="false"
        />
      </div>
    </label>
  </li>
  
</ul>
`;
